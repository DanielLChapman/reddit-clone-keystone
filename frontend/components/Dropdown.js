import React, { useState } from 'react';
import PropTypes from 'prop-types'

function Dropdown(props) {
    /*
        let options = ['Hot', 'Best', 'New', 'Top', 'Rising'];
        const [dropValue, setDropValue] = useState('Hot');
        <Dropdown returnFunc={setDropValue} selected={dropValue} options={options} />    
    */
    return (
        <select value={props.selected} onChange={(e) => {
            props.returnFunc(e.target.value);
        }}>
            {
                props.options && props.options.map((x, i) => (
                    <option key={i} value={x}>{x}</option>
                ))
            }
        </select>
    );
}

Dropdown.propTypes = {
    returnFunc: PropTypes.func.isRequired,
    options: PropTypes.array.isRequired,
    selected: PropTypes.string.isRequired,
};
  

export default Dropdown;