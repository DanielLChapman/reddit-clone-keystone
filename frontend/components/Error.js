import styled from 'styled-components';
import React from 'react';

import PropTypes from 'prop-types';

const ErrorStyles = styled.div`
`;

const Error = ({ error }) => {
  if (!error || !error.message) return null;
  if (error.networkError && error.networkError.result && error.networkError.result.errors.length) {
    return error.networkError.result.errors.map((error, i) => (
      <ErrorStyles key={i}>
        <p data-test="graphql-error">
          <strong>Error: </strong>
          {error.message.replace('GraphQL error: ', '')}
        </p>
      </ErrorStyles>
    ));
  }
  return (
    <ErrorStyles>
      <p data-test="graphql-error">
        <strong>Error: </strong>
        {error.message.replace('GraphQL error: ', '')}
      </p>
    </ErrorStyles>
  );
};

Error.defaultProps = {
  error: {},
};

Error.propTypes = {
  error: PropTypes.object,
};

export default Error;
