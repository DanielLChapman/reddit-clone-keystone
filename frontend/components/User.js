import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }
  }
  }
`;

export function useUser() {

  const { data } = useQuery(CURRENT_USER_QUERY);
  let test = {...data};

  
  if (test.authenticatedItem) {
    let x = Object.assign({}, test.authenticatedItem);
    //count up post votes
    let postVotes = 0;
    let commentVotes = 0;
    
    x.posts.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? postVotes += 1 : postVotes -= 1;
      })
    });
    //count up comment votes
    x.comments.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? commentVotes += 1 : commentVotes -= 1;
      })
    });

    x.postVotes = postVotes;
    x.commentVotes = commentVotes;

    return x;
  } else {
    return data?.authenticatedItem;
  }

  
}
