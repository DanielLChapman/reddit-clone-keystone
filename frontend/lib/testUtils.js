import casual from 'casual';

  
// seed it so we get consistent results
casual.seed(777);

const fakeUser = (overrides) => (
    {
    __typename: 'User',
    id: '4234',
    name: casual.name,
    username: casual.name,
    owner: [],
    email: casual.email,
    permissions: ['ADMIN'],
    moderating: [],
    subreddits: [],
    posts: [],
    postvotes: [],
    firstSignin: false,
    comments: [],
    commentvotes:[],
    ...overrides,
});

const fakeSubreddit = (overrides) => (
    {
        __typename: 'Subreddit',
        id: casual.uuid,
        name: casual.string,
        title: casual.string,
        slug: casual.string,
        sidebar: casual.string,
        rules: casual.string,
        description: casual.string,
        status: 'PUBLIC',
        ownder: fakeUser(),
        moderators: [],
        subscriber: [],
        posts: [],
        createdAt: Date.now(),
    }
);

const fakeLinkPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'www.google.com',
    type: 'Link',
    comments: [],
    removed: 'False',
    id: casual.uuid,


});

const fakeMediaPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'https://cdn.fileinfo.com/img/ss/sm/png_79.jpg',
    type: 'Link',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeTextPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    content: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: '',
    type: 'Text',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeComment = (overrides) => (
    {
    __typename: 'Comment',
    content: casual.sentence,
    user: fakeUser(),
    post: fakeTextPost(),
    createdAt: Date.now(),
    votes: [],
    parent: null,
    id: casual.uuid,
    children: [],

});


// Fake LocalStorage
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  }



export {

    LocalStorageMock,
    fakeLinkPost,
    fakeComment,
    fakeMediaPost,
    fakeSubreddit,
    fakeTextPost,
    fakeUser,

  };