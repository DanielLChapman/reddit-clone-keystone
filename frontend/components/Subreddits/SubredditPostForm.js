import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import React from 'react';
import useForm from '../../lib/useForm';
import Error from '../Error';
import Form from '../styles/SubredditSubmitForm';


const CREATE_VOTE = gql`
    mutation CREATE_VOTE($post_id: ID!, $vflag: String!) {
        createPostVote(data: {
            post: {
                connect: {
                    id: $post_id,
                }
            },
            vflag: $vflag
        }) {
            id
            vflag
        }
    }
`;


export const SUBREDDIT_TEXT_POST_MUTATION = gql`
    mutation SUBREDDIT_TEXT_POST_MUTATION($title: String!, $content: String,  $subreddit_id: ID!, $type: String!, $link: String ) {
        createVariedPost(
            title: $title,
            content: $content,
            subreddit_id: $subreddit_id,
            type: $type,
            link: $link,
        ) {
            id
            content
            post_slug
            title
        }
    }
`;

function sanitizeString(str){
    str = str.replace(/[^a-z0-9áéíóúñü \.,_-]/gim,"");
    return (str.length > 300) ? str.substr(0, 300-1) + '&hellip;' : str;
}

function SubredditPostForm(props) {
    const { inputs, handleChange, resetForm } = useForm({
        content: '',
        title: '',
      });


    const [createVariedPost, { data, error, loading}] = useMutation(
        SUBREDDIT_TEXT_POST_MUTATION
    );

    return (
        <div>
            {error && (
                <Error error={error} />
            )}
            {data && (
                <span>Success. <a href={`/r/${props.subreddit_slug}/comments/${data.createVariedPost.id}/${data.createVariedPost.post_slug}`}>Click here to view your new post!</a></span>
            )}
        <Form onSubmit={async (e) => {
            e.preventDefault();


            //NEED TO CREATE A POSTVOTE
            const res = await createVariedPost({
                variables: {
                    title: sanitizeString(inputs.title),
                    content: inputs.content,
                    subreddit_id: props.id,
                    link: '',
                    type: 'text'
                }
            });
            
            /* Moved to back end!! 
            const voteres = await createPostVote({
                variables: {
                    post_id: res.createVariedPost.id,
                    vflag: 'Upvote'
                }
            });

            */
            /*

            Router.push({
                pathname: `/r/${inputs.name.toLowerCase()}`,
              });
*/
        }}>
            <fieldset>
                    <label htmlFor="title">
                        Title <br />
                        <input 
                            required
                            type="text"
                            id="title"
                            name="title"
                            value={inputs.title}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="content">
                        Text: <br />
                        <textarea
                            id="content"
                            name="content"
                            placeholder=""
                            value={inputs.content}
                            onChange={handleChange}
                        />
                    </label>
                    <button type="submit">Create Text Post</button>
                    
                </fieldset>



        </Form>
        </div>

    );
}

export default SubredditPostForm;