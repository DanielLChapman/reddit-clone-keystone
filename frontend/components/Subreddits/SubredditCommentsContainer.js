import React from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { FaLongArrowAltUp, FaLongArrowAltDown } from 'react-icons/fa';
import { useMutation, useQuery } from '@apollo/client';
import { CREATE_VOTE, DELETE_VOTE, UPDATE_VOTE } from '../SmallPosts/PostMain';
import { CURRENT_USER_QUERY, useUser } from '../User';
import CommentsPage from './SubredditComments/CommentsPage';
import SubredditTopBar from './SubredditComments/CommentsPageTopBar';

const GET_POST_INFO = gql`
    query GET_POST_INFO($id: ID!) {
        Post( where: { id: $id } )
    {
        id
        content
        title
        user {
            id
            name
            username
        }
        createdAt
        votes {
            vflag
        }
        post_slug
        link
        removed
        subreddit {
            slug
        }
        type
        comments {
            id
            parent {
                id
            }
            content
            createdAt
            user {
                id
                username
            }
            votes {
                vflag
            }
        }

    }
    }
`;

const GET_SPECIFIC_COMMENTS = gql`
    query GET_SPECIFIC_COMMENTS($id: ID!) {
        Comment (
            where: {
                id: $id
            }
        ) {
            id
            parent {
                id
            }
            content
            createdAt
            user {
                id
                username
            }
            votes {
                vflag
            }
            children {
                id
                parent {
                    id
                }
                content
                createdAt
                user {
                    id
                    username
                }
                votes {
                    vflag
                }
                children {
                    id
                    parent {
                        id
                    }
                    content
                    createdAt
                    user {
                        id
                        username
                    }
                    votes {
                        vflag
                    }
                    children {
                        id
                        parent {
                            id
                        }
                        content
                        createdAt
                        user {
                            id
                            username
                        }
                        votes {
                            vflag
                        }
                        children {
                            id
                            parent {
                                id
                            }
                            content
                            createdAt
                            user {
                                id
                                username
                            }
                            votes {
                                vflag
                            }
                        }
                    }
                }
            }
        } 
    }
`;

function SubredditCommentsContainer(props) {
    const user = useUser();

    const {data, error, loading} = useQuery(GET_POST_INFO, {
        variables: {
            id: props.postid,
        }
    });

    

    if (error) return <span>Error...</span>
    if (loading) return <span>Loading...</span>

    let post = {...data?.Post};
    let style = {
        display: 'block'
    }
    let showLinkToGoBackToAllComments = false;

    if (props.commentArray) {
        post.comments = props.commentArray;
        showLinkToGoBackToAllComments = true;
        style = {
            display: 'none'
        }
    }

    return (
        <div className="comment-page">
            <SubredditTopBar user={user} post={post} subreddit={props.subreddit}/>
            <section className="comment-container" >
                
                <CommentsPage showLinkToGoBackToAllComments={showLinkToGoBackToAllComments} propstyle={style} ownership={props.ownership} user={user} post={post} subreddit={props.subreddit} />
            </section>
        </div>
    );
}

// SubredditCommentsContainer.propTypes = {
//     subreddit: PropTypes.object.isRequired,
//     postid: PropTypes.string.isRequired,

// }

export default SubredditCommentsContainer;