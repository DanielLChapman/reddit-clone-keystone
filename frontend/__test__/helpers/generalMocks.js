import { CURRENT_USER_QUERY } from '../../components/User';
import { fakeTextPost, fakeUser} from '../../lib/testUtils';

const fakeUserObj = fakeUser();
const notSignedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: null } },
    },
  ];
  
const signedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: fakeUser() /*{
        
            id: fakeUserObj.id,
            username: fakeUserObj.username,
            email: fakeUserObj.email,
            name:fakeUserObj.name,
            posts: [],
            comments: [],
            moderating: [],
            owner:[],
            subreddits:[],
            postvotes:[],
            commentvotes: [],
        
      }*/ } },
    },
  ];
  
const signedInMocksWithPosts = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: {
        data: {
          authenticatedItem: fakeUser({
            posts: [fakeTextPost()],
          }),
        },
      },
    },
];

export {
    notSignedInMocks,
    signedInMocks,
    signedInMocksWithPosts
}

/* 

{
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }

    */