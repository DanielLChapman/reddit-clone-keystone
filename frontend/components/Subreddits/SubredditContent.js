import Link from 'next/link';
import React from 'react';

import PropTypes from 'prop-types';
import SubredditPostForm from './SubredditPostForm';
import SubredditHomePage from './SubredditHomePage';
import SubredditCommentsContainer from './SubredditCommentsContainer';
import SubredditMediaForm from './SubredditMediaForm';
import { faIgloo } from '@fortawesome/free-solid-svg-icons';
import { useUser } from '../User';
import checkUserToSubreddit from '../../lib/checkUser';


function SubredditContent(props) {
    let user = useUser();

    let ownership = checkUserToSubreddit(user, props.subreddit.moderators, props.subreddit.owner);

    switch(props.type) {
        case 'Home': 
            if (props.slug && props.subreddit) {
                return <SubredditHomePage ownership={ownership || false} slug={props.slug} id={props.subreddit.id} subreddit={props.subreddit}/>
            }
            return <span>Something was missing here</span>
        case 'comment':
            if (props.slug && props.subreddit && props.postslug && props.postid) {
                return <SubredditCommentsContainer ownership={ownership || false} slug={props.slug} subreddit={props.subreddit} postslug={props.postslug} postid={props.postid}/>
            }
            return <span>Something was missing here</span>
        case 'permalinkedcomments':
            if (props.slug && props.subreddit && props.postslug && props.postid && props.commentid && props.commentArray) {
                return <SubredditCommentsContainer commentid={props.commentid} commentArray={props.commentArray} ownership={ownership || false} slug={props.slug} subreddit={props.subreddit} postslug={props.postslug} postid={props.postid}/>
            }
            return <span>Something was missing here!</span>
        case 'form':
            if (props.selftext && props.selftext === "true") {
                return (
                    <SubredditPostForm slug={props.slug} id={props.subreddit.id} />
                )
            }
            return (
                <SubredditMediaForm slug={props.slug} id={props.subreddit.id} />
            )
        default: 
            return (
                <span>Something went wrong...</span>
            )
    }
    

}


export default SubredditContent;