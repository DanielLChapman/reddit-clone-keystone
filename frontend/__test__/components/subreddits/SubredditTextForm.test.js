import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import SubredditContent from '../../../components/Subreddits/SubredditContent';
import { fakeSubreddit, fakeTextPost, fakeUser  } from '../../../lib/testUtils';
import wait from 'waait';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';
import { GET_POST_INFO } from '../../../components/Subreddits/SubredditCommentsContainer';
//import Signup, { SIGNUP_MUTATION } from '../../components/SignUp';
import { signedInMocksWithPosts, signedInMocks } from '../../helpers/generalMocks';
import { CURRENT_USER_QUERY } from '../../../components/User';
import { SUBREDDIT_TEXT_POST_MUTATION } from '../../../components/Subreddits/SubredditPostForm';

const fakeSubredditObj = fakeSubreddit();
const fakePostObj = fakeTextPost();
const fakeUserObj = fakeUser();



const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { authenticatedItem: {...fakeUserObj}}}
    },
    {
        request: {
            query: SUBREDDIT_TEXT_POST_MUTATION,
            variables: {
                title:fakePostObj.title,
                content:fakePostObj.content,
                subreddit_id:fakePostObj.subreddit.id,
                link:"",
                type:"text"}
        },
        result: {
            data: {
                createVariedPost: {
                    id: fakePostObj.id,
                    content: fakePostObj.content,
                    post_slug: fakePostObj.post_slug,
                    title: fakePostObj.title,
                }
            }
        }

    }

  ];

describe('It switches subreddit content based on props', () => {
    it ('tries to renders a text form with being logged in', async () => {
        const { container, debug} = render(
            <MockedProvider mocks={mocks} >
                <SubredditContent type="form" selftext="true" id={fakePostObj.subreddit.id} subreddit={fakePostObj.subreddit}/>
            </MockedProvider>
         )
       

        await screen.findByText('Create Text Post'); 

        expect(container).toMatchSnapshot();
    });

    it ('succesfully submits data and gets a desired response', async () => {
        const { container, debug} = render(
            <MockedProvider mocks={mocks} >
                <SubredditContent type="form" selftext="true" id={fakePostObj.subreddit.id}  subreddit={fakePostObj.subreddit}/>
            </MockedProvider>
         )
       

        await screen.findByText('Create Text Post'); 

        await userEvent.type(screen.getByLabelText('Title'), fakePostObj.title);
        await userEvent.type(screen.getByLabelText('Text:'), fakePostObj.content);
        await userEvent.click(screen.getByText('Create Text Post'));

        await waitFor(() => wait(0));

        await screen.findByText('Success.'); 
    })


})