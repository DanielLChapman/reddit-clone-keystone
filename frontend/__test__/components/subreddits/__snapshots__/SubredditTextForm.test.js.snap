// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It switches subreddit content based on props tries to renders a text form with being logged in 1`] = `
<div>
  <div>
    <form
      class="SubredditSubmitForm__Form-sc-4m25ol-0 dSAjAw"
    >
      <fieldset>
        <label
          for="title"
        >
          Title 
          <br />
          <input
            id="title"
            name="title"
            required=""
            type="text"
            value=""
          />
        </label>
        <label
          for="content"
        >
          Text: 
          <br />
          <textarea
            id="content"
            name="content"
            placeholder=""
          />
        </label>
        <button
          type="submit"
        >
          Create Text Post
        </button>
      </fieldset>
    </form>
  </div>
</div>
`;
